src/ORM/Hierarchy/Hierarchy.php

```diff
@@ -407,25 +409,47 @@ public function showingCMSTree()
             && in_array($controller->getAction(), ["treeview", "listview", "getsubtree"]);
     }
 
+    /**
+     * Find the first class in the inheritance chain that has Hierarchy extension applied
+     *
+     * @return string
+     */
+    private function getHierarchyBaseClass(): string
+    {
+        $ancestry = ClassInfo::ancestry($this->owner);
+        $ancestorClass = array_shift($ancestry);
+        while ($ancestorClass && !Extensible::has_extension($ancestorClass, self::class)) {
```

Comment:
Missed opportunity for a `do..while` 🙁

Reply:
I think I tried that first but either I couldn't make it work or it look worse for some reason, but I started with it and then went back to the `while` loop. Can have another look and raise a tiny enhancement PR at some stage.